### ISSUE TITLE
[Question] How to do multi-task learning 

### ISSUE THREAD

**Issue by shoyasaxa** [NVIDIA-Merlin/HugeCTR#290]
Hello, 

I am unable to get multi-task learning with multiple binary labels working, and I was wondering if you could help me figure out what is wrong. 

In this synthetic example below, I create a simple dataset with 3 possible values for each label, preprocess it, and attempt to train a simple DLRM on it. However, I get an error upon trying to define the model. 

```
# Data Generation 

import numpy as np
import pandas as pd
import random 

num_sampled = 10000

labels = random.choices(
    [[1.0,0.0,0.0],
    [0.0,1.0,0.0],
    [0.0,0.0,1.0]], 
    k=num_sampled
)
df = pd.DataFrame()
df["label"] = labels

cat_cols = []
for i in range(4):
    cat_col = f"cat_{i}"
    cat_vals = random.choices(list(range(10)),k=num_sampled)
    df[cat_col] = cat_vals
    cat_cols.append(cat_col)

cont_cols = [] 
for i in range(3):
    cont_col = f"cont_{i}"
    cont_vals = np.random.uniform(0,1,size=num_sampled)
    df[cont_col] = cont_vals
    cont_cols.append(cont_col)

# df 

| label | cat_0 | cat_1 | cat_2 | cat_3 | cont_0 | cont_1 | cont_2
-- | -- | -- | -- | -- | -- | -- | -- | --
[1.0, 0.0, 0.0] | 9 | 5 | 2 | 0 | 0.788101 | 0.250695 | 0.946659
[0.0, 0.0, 1.0] | 0 | 9 | 4 | 0 | 0.222429 | 0.543690 | 0.960873
[0.0, 1.0, 0.0] | 4 | 5 | 6 | 3 | 0.273892 | 0.170089 | 0.142081
[1.0, 0.0, 0.0] | 9 | 7 | 7 | 5 | 0.859505 | 0.358539 | 0.072872
[0.0, 1.0, 0.0] | 7 | 0 | 5 | 0 | 0.211338 | 0.566059 | 0.816855
... | ... | ... | ... | ... | ... | ... | ...
[0.0, 1.0, 0.0] | 1 | 2 | 2 | 2 | 0.735870 | 0.975657 | 0.828746
[0.0, 0.0, 1.0] | 6 | 0 | 7 | 3 | 0.334677 | 0.850328 | 0.845184
[0.0, 0.0, 1.0] | 5 | 4 | 6 | 3 | 0.657202 | 0.770733 | 0.781514
[1.0, 0.0, 0.0] | 1 | 6 | 6 | 2 | 0.133572 | 0.622814 | 0.121681
[0.0, 0.0, 1.0] | 4 | 6 | 5 | 5 | 0.578604 | 0.085959 | 0.885711

from sklearn.model_selection import train_test_split

train, test = train_test_split(df, test_size=0.2)

# Preprocessing
import nvtabular as nvt 
from nvtabular.ops import Categorify, Normalize, FillMissing, Rename, FillMedian

cat_cols_processed = nvt.ColumnGroup(cat_cols) >> Categorify()
cont_cols_processed = nvt.ColumnGroup(cont_cols) >> Normalize() >> FillMissing() 
output = cat_cols_processed + cont_cols_processed + ["label"]
workflow = nvt.Workflow(output)

dict_dtypes = {}
for col in cat_cols:
    dict_dtypes[col] = np.int64
for col in cont_cols:
    dict_dtypes[col] = np.float32

# dict_dtypes 
{'cat_0': numpy.int64,
 'cat_1': numpy.int64,
 'cat_2': numpy.int64,
 'cat_3': numpy.int64,
 'cont_0': numpy.float32,
 'cont_1': numpy.float32,
 'cont_2': numpy.float32}

train_dataset = nvt.Dataset(train)
test_dataset = nvt.Dataset(test)

workflow.fit(train_dataset)

workflow.transform(train_dataset).to_parquet(
    output_path="/usr/local/mnt/data/multi_task/data/train",
    out_files_per_proc=1,
    dtypes=dict_dtypes,
    cats=cat_cols,
    conts=cont_cols,
    labels=["label"]
)
workflow.transform(test_dataset).to_parquet(
    output_path="/usr/local/mnt/data/multi_task/data/test",
    out_files_per_proc=1,
    dtypes=dict_dtypes,
    cats=cat_cols,
    conts=cont_cols,
    labels=["label"]
)

# Model Training 
import hugectr
from mpi4py import MPI

solver = hugectr.CreateSolver(
    max_eval_batches = 32,
    batchsize_eval = 100,
    batchsize = 128,
    lr = 0.01,
    end_lr = 0.0001,
    warmup_steps = 8000,
    decay_start = 48000,
    decay_steps = 24000,
    vvgpu = [[0]],
    repeat_dataset = True,
    i64_input_key = True
)

reader = hugectr.DataReaderParams(
    data_reader_type = hugectr.DataReaderType_t.Parquet,
    source = ["/usr/local/mnt/data/multi_task/data/train/_file_list.txt"],
    eval_source = "/usr/local/mnt/data/multi_task/data/test/_file_list.txt",
    check_type = hugectr.Check_t.Sum,
    slot_size_array = [11, 11, 11, 11]
)

optimizer = hugectr.CreateOptimizer(
    optimizer_type = hugectr.Optimizer_t.Adam,
    update_type = hugectr.Update_t.Local,
    beta1 = 0.9,
    beta2 = 0.999,
    epsilon = 0.0000001
)

model = hugectr.Model(solver, reader, optimizer)
# output 
HugeCTR Version: 3.2
====================================================Model Init=====================================================
[HUGECTR][16:59:06][INFO][RANK0]: Global seed is 2655792394
[HUGECTR][16:59:07][INFO][RANK0]: Device to NUMA mapping:
  GPU 0 ->  node 0

[HUGECTR][16:59:08][WARNING][RANK0]: Peer-to-peer access cannot be fully enabled.
[HUGECTR][16:59:08][INFO][RANK0]: Start all2all warmup
[HUGECTR][16:59:08][INFO][RANK0]: End all2all warmup
[HUGECTR][16:59:08][INFO][RANK0]: Using All-reduce algorithm: NCCL
[HUGECTR][16:59:08][INFO][RANK0]: Device 0: Tesla V100-SXM2-16GB
####

# Define a simple DLRM architecture
model.add(
    hugectr.Input(
        label_dim=3,
        label_name="label",
        dense_dim=3,
        dense_name="dense",
        data_reader_sparse_param_array=[
            hugectr.DataReaderSparseParam(
                "data1",
                [1,1,1,1],
                False,
                4,
            )
        ],
    )
)

model.add(
    hugectr.SparseEmbedding(
        embedding_type=hugectr.Embedding_t.LocalizedSlotSparseEmbeddingHash,
        workspace_size_per_gpu_in_mb=10,
        embedding_vec_size=16,
        combiner="mean",
        sparse_embedding_name="sparse_embedding1",
        bottom_name="data1",
        optimizer=optimizer,
    )
)
model.add(
    hugectr.DenseLayer(
        layer_type=hugectr.Layer_t.InnerProduct,
        bottom_names=["dense"],
        top_names=["fc1"],
        num_output=16,
    )
)
model.add(
    hugectr.DenseLayer(
        layer_type=hugectr.Layer_t.ReLU,
        bottom_names=["fc1"],
        top_names=["relu1"],
    )
)
model.add(
    hugectr.DenseLayer(
        layer_type=hugectr.Layer_t.Interaction,
        bottom_names=["relu1", "sparse_embedding1"],
        top_names=["interaction1"],
    )
)
model.add(
    hugectr.DenseLayer(
        layer_type=hugectr.Layer_t.InnerProduct,
        bottom_names=["interaction1"],
        top_names=["fc4"],
        num_output=32,
    )
)
model.add(
    hugectr.DenseLayer(
        layer_type=hugectr.Layer_t.ReLU,
        bottom_names=["fc4"],
        top_names=["relu4"],
    )
)
model.add(
    hugectr.DenseLayer(
        layer_type=hugectr.Layer_t.InnerProduct,
        bottom_names=["relu4"],
        top_names=["fc8"],
        num_output=3,
    )
)
model.add(
    hugectr.DenseLayer(
        layer_type=hugectr.Layer_t.MultiCrossEntropyLoss,
        bottom_names=["fc8", "label"],
        top_names=["loss"],
        target_weight_vec=[0.2,0.4,0.4]
    )
)
model.compile()

```
This `model.compile()`outputs
```
Runtime error: Parquet reader: Dense KeyType and Parquet column type don't match /var/tmp/HugeCTR/HugeCTR/include/data_readers/parquet_data_reader_worker.hpp:362 
Runtime error: Parquet reader: Dense KeyType and Parquet column type don't match /var/tmp/HugeCTR/HugeCTR/include/data_readers/parquet_data_reader_worker.hpp:362 


[HUGECTR][16:59:08][INFO][RANK0]: Graph analysis to resolve tensor dependency
===================================================Model Compile===================================================
[HUGECTR][16:59:10][INFO][RANK0]: gpu0 start to init embedding
[HUGECTR][16:59:10][INFO][RANK0]: gpu0 init embedding done
[HUGECTR][16:59:10][INFO][RANK0]: Starting AUC NCCL warm-up
[HUGECTR][16:59:10][INFO][RANK0]: Warm-up done
```
If I execute `model.summary()` after this, I get the correct output 
```
===================================================Model Summary===================================================
label                                   Dense                         Sparse                        
label                                   dense                          data1                         
(None, 3)                               (None, 3)                               
------------------------------------------------------------------------------------------------------------------
Layer Type                              Input Name                    Output Name                   Output Shape                  
------------------------------------------------------------------------------------------------------------------
LocalizedSlotSparseEmbeddingHash        data1                         sparse_embedding1             (None, 4, 16)                 
InnerProduct                            dense                         fc1                           (None, 16)                    
ReLU                                    fc1                           relu1                         (None, 16)                    
Interaction                             relu1,sparse_embedding1       interaction1                  (None, 27)                    
InnerProduct                            interaction1                  fc4                           (None, 32)                    
ReLU                                    fc4                           relu4                         (None, 32)                    
InnerProduct                            relu4                         fc8                           (None, 3)                     
MultiCrossEntropyLoss                   fc8,label                     loss                                                        
------------------------------------------------------------------------------------------------------------------
```

I suspect this is error springs from the fact that I do not pass in the `label` column's save data type in NVT (the `dict_dtypes` in the code above). However, I cannot set the save dtype as `float32` because the label column is an array type (since we are doing multi-task learning), and NVT does not explicitly support saving as `List[float32]`. So, I instead opted to set the value of the label column as list of floats "manually" in the data generation, but that did not work. 

Also another question is - is setting the `label` column as an array column correct? Or do I need multiple different label columns (i.e. `label1`, `label2`, `label3`)?

**Comment by shoyasaxa** [comment]
I was able to get it to work by not having a single array column and separating the label columns into their own columns
